### PR TITLE
resolve, `try_resolve_as_non_binding`: use `delay_span_bug` due to parser recovery

### LINKED ISSUE #70549: Internal Compiler Error: unexpected resolution for an identifier in pattern: SelfCtor(DefId(0:4))
I'm seeing an internal compiler error on the following input (found by [fuzz-rustc](https://github.com/dwrensha/fuzz-rustc)):

```rust
struct S {}
impl S {
    fn foo(&mur Self) {}
}
fn main() {}
```

```
$ rustc main.rs 
error: expected identifier, found keyword `Self`
 --> main.rs:3:17
  |
3 |     fn foo(&mur Self) {}
  |                 ^^^^ expected identifier, found keyword

error: expected one of `:`, `@`, or `|`, found keyword `Self`
 --> main.rs:3:17
  |
3 |     fn foo(&mur Self) {}
  |            -----^^^^
  |            |    |
  |            |    expected one of `:`, `@`, or `|`
  |            help: declare the type after the parameter binding: `<identifier>: <type>`

thread 'rustc' panicked at 'src/librustc_resolve/late.rs:1572: unexpected resolution for an identifier in pattern: SelfCtor(DefId(0:4))', /rustc/699f83f525c985000c1f70bf85117ba383adde87/src/libstd/macros.rs:13:23
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.44.0-nightly (699f83f52 2020-03-29) running on x86_64-unknown-linux-gnu

error: aborting due to 2 previous errors
```

The error happens on nightly but not on beta or stable.

### PR DESCRIPTION
**Pull request by Centril**
Fixes #70549

r? @petrochenkov

### CODE REVIEW

File: src/librustc_resolve/late.rs [hunk 1]
```diff
@@ -1536,20 +1536,17 @@ impl<'a, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
         let is_syntactic_ambiguity = !has_sub && bm == BindingMode::ByValue(Mutability::Not);
 
         match res {
-            Res::Def(DefKind::Ctor(_, CtorKind::Const), _)
-            | Res::Def(DefKind::Const, _)
-            | Res::Def(DefKind::ConstParam, _)
-                if is_syntactic_ambiguity =>
-            {
+            Res::Def(
+                DefKind::Ctor(_, CtorKind::Const) | DefKind::Const | DefKind::ConstParam,
```
Comment (petrochenkov) [inline src/librustc_resolve/late.rs]: Could you add `Res::SelfCtor` to this list instead?
`Self` is an entirely reasonable const pattern (https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8eea5be549cc8d5ae53354d21ab90d3c), and the fact that it's generally interpreted as an `PatKind::Path` and doesn't appear here is a parser's implementation detail.

Comment (petrochenkov) [reply src/librustc_resolve/late.rs]: Also, `delay_span_bug` is a hack and I never use it in code I maintain.

Comment (Centril) [reply src/librustc_resolve/late.rs]: Done. (It does regress diagnostics a bit.)

**Pull request merged**